Adding non critical demand sink

Bachibouzouk opened the pull request.

adnanalakori committed:
Adding non critical demand sink

Bachibouzouk commented on critical_demand.py:
```diff
@@ -127,15 +127,15 @@
                 nominal_value=None,
                 investment=solph.Investment(
-                    ep_costs=epc_pv * n_days / n_days_in_year
+                    ep_costs=epc_pv * n_days / n_days_in_year #ADN:why not just put ep_costs=epc_PV??
```
Maybe we should modify epc_pv by the factor `(n_days / n_days_in_year)` only in the results

Bachibouzouk commented on critical_demand.py:
```diff
@@ -226,16 +228,15 @@
         outflow_conversion_factor=0.9,
         invest_relation_input_capacity=1,
-        invest_relation_output_capacity=0.5,
+        invest_relation_output_capacity=0.5, #fixes the input flow investment to the output flow investment #ADN:why 0.5?
```
@adnanalakori - can you provide a link to the documentation there?

adnanalakori replied:
https://oemof-solph.readthedocs.io/en/latest/usage.html#using-an-investment-object-with-the-genericstorage-component

Bachibouzouk commented on critical_demand.py:
```diff
@@ -321,22 +330,23 @@
 results_rectifier = solph.views.node(results=results, node="rectifier")
 if case != case_BPV:
-    results_battery = solph.views.node(results=results, node="battery")
+    results_battery = solph.views.node(results=results, node="diesel")
```
```suggestion
    results_diesel = solph.views.node(results=results, node="diesel")
```

Bachibouzouk replied:
moreover you should provide `results_battery` for the case `case != CASE_D`

Bachibouzouk commented on critical_demand.py:
```diff
@@ -346,7 +356,7 @@
+# liter by dividing it by the diesel density.#ADN:??
 sequences_diesel_consumption = (
     results_diesel_source["sequences"][(("diesel_source", "diesel"), "flow")]
     / diesel_lhv
```
don't divide by any factor if you want kWh for diesel

Bachibouzouk committed:
Fix typo

Bachibouzouk committed:
Make usecase run

adnanalakori committed:
add jupter

Bachibouzouk committed:
Make the model close to initial case

Still separating between diesel, battery+PV, battery+diesel+PV

Bachibouzouk committed:
Adapt total demand calculation

Bachibouzouk committed:
Lint code to make it pretty

Bachibouzouk committed:
Add dash app to visualize the results

Bachibouzouk committed:
Merge branch 'feature/results-display' into feature/working-model

adnanalakori commented on critical_demand.py:
```diff
@@ -239,24 +241,27 @@
-            fix=non_critical_demand, #/ non_critical_demand.max(),
-            #nominal_value=1, #non_critical_demand.max(),
+            min=(1-demand_reduction_factor)* (non_critical_demand/ non_critical_demand.max()),
+            max=(non_critical_demand/ non_critical_demand.max()),
```
It is not clear to me why we also not multiply `max `  with the `1-demand_reduction_factor`

Bachibouzouk replied:
if you can pick a value between min and max and min=max, then you don't choose a value, it has to be min, or max (because they are then same)

Bachibouzouk replied:
(this would be equivalent to using fix)

Bachibouzouk committed:
Separate model optimization and results post-processing

Bachibouzouk committed:
Delete commented unused code

Bachibouzouk committed:
Add function to plot demand reduction

Bachibouzouk committed:
Remove unused imports

Bachibouzouk committed:
Work on app layout

Bachibouzouk committed:
Add a dropdown menu to select a timeslice

Bachibouzouk committed:
Add template case input file

Bachibouzouk committed:
Add xlsx files to gitignore

Bachibouzouk committed:
Update requirements.txt

Bachibouzouk committed:
Add inputfile reader

Bachibouzouk committed:
Read settings costs and timeseries from input file

Bachibouzouk committed:
Reassamble optimization and results processing into one function

Dash does not allow optimization in callbacks anyway

Bachibouzouk committed:
Add black line in demand_supplied graph

Bachibouzouk committed:
Allow user to change port of dash app

Bachibouzouk committed:
Included new columns in input file

Bachibouzouk committed:
Add graph of energy system if oemof.visio is installed

Bachibouzouk committed:
Enable input of file with -i option

python critical_demand.py -i <name of the case file>

Bachibouzouk committed:
Implement calculation of KPIs

Bachibouzouk committed:
Replace epc by annuity

Bachibouzouk committed:
Comment out linear depreciation of component

Bachibouzouk committed:
Move case settings into the function call

Bachibouzouk committed:
Add input parameters and output to the dash app

Bachibouzouk committed:
Update template file

Bachibouzouk committed:
Move executed code into __main__

Bachibouzouk committed:
Add parser to sensitivity inputs

Bachibouzouk committed:
Script to perform sensitivity analysis

Bachibouzouk committed:
Store one simulation results into a df

Making difference between asset/component results (optimized capacities,
total flow, etc) and system results (KPI, total costs, etc)

Bachibouzouk committed:
Collect the results of sensitivity analysis

Bachibouzouk committed:
Rename results variable with better name

Bachibouzouk committed:
Move the sensitivity code into a function

Bachibouzouk committed:
Add a dash app for looking at sensitivity results

Bachibouzouk committed:
Change the collection of sensitivity analysis results

Encompass assets and kpi results as function of input parameters
from both settings and assets

Bachibouzouk committed:
Improve the dash app to be able to plot all x versus all y

adnanalakori commented on critical_demand.py:
```diff
@@ -346,6 +348,16 @@ def run_simulation(n_days=n_days, case=case):
+
+    project_lifetime = 20
+    wacc = 0.06
+    CRF = annuity(1, project_lifetime, wacc)
```
Do we need to add annuity here? to calculate CRF, we can apply this equ.:  
CRF=(wacc ⋅ (1+wacc)^T)/(〖(1+wacc)〗^T-1)

Bachibouzouk replied:
Annuity with a capex of 1 is equal to CRF : https://oemof-solph.readthedocs.io/en/v0.2.3/api/oemof.tools.html#oemof.tools.economics.annuity

adnanalakori commented on critical_demand.py:
```diff
@@ -35,513 +35,892 @@
+    ]
+    df_results.to_csv(f"results_{case}.csv")
+
+    NPV = (df_results.annual_costs.sum() + df_results.cash_flow.sum()) / CRF
```
Hi @Bachibouzouk , I guess `annual_costs` and `cash_flow` are not defined yet. Also, another important KPI which is the `first_investment` for the three cases is also missed.

Bachibouzouk replied:
@adnanalakori  If you can run the code without error that means that `annual_costs` and `cash_flow` are defined somewhere above this line. Use `ctrl + f` to look occurences of `annual_costs` (name of the column in the dataframe `df_results`) or `cash_flow`

adnanalakori replied:
Thanks, I figured its value through looking for whole variable  `df_results.annual_costs.sum`, before I was debugging  `annual_costs` or `cash_flow` which shows as an empty (non defined) string one.

Bachibouzouk commented on critical_demand.py:
```diff
@@ -35,513 +35,892 @@
-##########################################################################
-# Process the results
-##########################################################################
+        df_results.loc["diesel_genset", "cash_flow"] = (
```
@adnanalakori This is the definition of cash_flow for diesel, you can look in the file where "cash_flow" is used (using ctrl+f function to search for occurences)

Bachibouzouk committed:
Change definition of options in dash app

Bachibouzouk committed:
Lint with black latest version

Bachibouzouk committed:
Merge pull request #3 from Bachibouzouk/feature/sensitivity

Sensitivity analysis

adnanalakori commented on utils.py:
```diff
@@ -0,0 +1,190 @@
+
+
+def annualized_capex(
+    investment_t0, project_lifetime, asset_lifetime=None, wacc=0.05, tax=0
```
To which  `investment_t0` refers to?

Bachibouzouk replied:
look in the code `capex_from_investment` function's docstring, alternatively you can also look where `capex_from_investment` is used in the code with ctrl+f

adnanalakori replied:
I looked before, it looks like it is added to the `capex_from_investment` but not defined.

Bachibouzouk replied:
you can also look where capex_from_investment is used in the code with ctrl+f

Bachibouzouk committed:
Add possibility to change the min and max soc level

Bachibouzouk committed:
Add the first invesment and expenditure fuel total in results

Bachibouzouk committed:
Fix display of first investment and add tooltip for lcoe

Bachibouzouk committed:
Change battery input parameters

Bachibouzouk committed:
Improve plotting of battery state of charge

Bachibouzouk committed:
Modify the definition of first investment

adnanalakori committed:
Adjust Peak demand and adjust NPV

Bachibouzouk commented on critical_demand.py:
```diff
@@ -496,7 +494,7 @@ def run_simulation(df_costs, data, settings):
     asset_results.to_csv(f"results_{case}.csv")
 
-    NPV = (asset_results.annual_costs.sum() + asset_results.cash_flow.sum()) / CRF
+    NPV = ((asset_results.annual_costs.sum() + asset_results.cash_flow.sum()) / CRF) + project_planning_cost
```
it is possible to define it in the input file, so that the user does not need to change this value in the .py code

adnanalakori replied:
Yes, we could add it later to the input file and let the user define it

Bachibouzouk commented on critical_demand.py:
```diff
@@ -541,6 +539,7 @@ def run_simulation(df_costs, data, settings):
     ) + non_critical_demand[sequences_demand.index].sum(axis=0)
 
     total_opex_costs = asset_results.total_opex_costs.sum() * project_lifetime
+    overall_peak_demand = sequences_demand.max() + sequences_critical_demand.max()
```
if you sum the max of the 2 sequences it is likely to be larger than the max of the sum

adnanalakori replied:
That is true.

adnanalakori committed:
Adding diesel genset's reserve factor

adnanalakori committed:
Adding diesel genset's reserve factor

adnanalakori committed:
Adding project planning cost to the first_investment

adnanalakori committed:
Opex per year instead of over project lifetime

adnanalakori committed:
Adding cash flow to the opex

adnanalakori committed:
Adding critical and non critical load to the results
